# Semantica y Vectores de palabra con Spacy

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Ohtar10/icesi-nlp/blob/main/Sesion2/1-semantics-word-vectors.ipynb)

En este notebook exploraremos el concepto de word vectors utilizando la librería Spacy.

## Referencias
* [NLP - Natural Language Processing With Python](https://www.udemy.com/course/nlp-natural-language-processing-with-python)

In [1]:
import pkg_resources
import warnings

warnings.filterwarnings('ignore')

installed_packages = [package.key for package in pkg_resources.working_set]
IN_COLAB = 'google-colab' in installed_packages

In [2]:
#!test '{IN_COLAB}' = 'True' && wget  https://github.com/Ohtar10/icesi-nlp/raw/refs/heads/main/requirements.txt && pip install -r requirements.txt

In [3]:
#!python -m spacy download en_core_web_lg

In [ ]:
#%pip install https://github.com/explosion/spacy-models/releases/download/en_core_web_lg-3.7.1/en_core_web_lg-3.7.1-py3-none-any.whl

import spacy

nlp = spacy.load('en_core_web_lg')

     ---------------------------------------- 0.0/587.7 MB ? eta -:--:--
     ---------------------------------------- 1.0/587.7 MB 6.3 MB/s eta 0:01:34
      -------------------------------------- 8.4/587.7 MB 23.7 MB/s eta 0:00:25
      ------------------------------------- 11.3/587.7 MB 20.7 MB/s eta 0:00:28
     - ------------------------------------ 19.1/587.7 MB 25.2 MB/s eta 0:00:23
     - ------------------------------------ 25.7/587.7 MB 26.3 MB/s eta 0:00:22
     -- ----------------------------------- 31.5/587.7 MB 28.9 MB/s eta 0:00:20
     -- ----------------------------------- 32.5/587.7 MB 23.2 MB/s eta 0:00:24
     -- ----------------------------------- 40.4/587.7 MB 25.2 MB/s eta 0:00:22
     --- ---------------------------------- 49.3/587.7 MB 27.1 MB/s eta 0:00:20
     --- ---------------------------------- 52.7/587.7 MB 26.2 MB/s eta 0:00:21
     --- ---------------------------------- 56.4/587.7 MB 24.9 MB/s eta 0:00:22
     ---- --------------------------------- 63.


[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


En esta ocasión, este modelo de Spacy más grande ya contiene vectores de palabras densos de 300 dimensiones que podemos utilizar de inmediato.

Por ejemplo, exploremos la palabra `lion`

In [8]:
nlp(u'lion').vector

array([  1.2746  ,   0.46242 ,  -1.1829  ,  -5.2661  ,  -2.7128  ,
         1.8521  ,  -0.94273 ,   2.1865  ,   6.503   ,   0.6704  ,
         1.5361  ,   2.5992  ,  -0.36233 ,   4.3965  ,  -6.5644  ,
         1.6141  ,  -1.2897  ,   2.1184  ,  -0.63654 ,  -3.4572  ,
        -4.3771  ,   4.2074  ,  -3.6411  ,  -0.97214 ,   1.3253  ,
        -2.3125  ,  -3.6531  ,  -2.8398  ,   2.7913  ,  -1.53    ,
        -2.9984  ,  -2.6357  ,   0.50615 ,  -2.6925  ,   4.3401  ,
        -5.6017  ,   0.045691,   4.3832  ,  -0.19535 ,  -1.0751  ,
         0.32172 ,   2.4395  ,   4.6638  ,   3.4471  ,  -3.3847  ,
        -1.8238  ,   0.70212 ,   0.58557 ,   5.0032  ,  -3.1072  ,
         1.2364  ,   7.4595  ,   0.057368,   1.0111  ,  -1.0827  ,
         0.69113 ,   2.8009  ,  -3.4383  ,  -1.0599  ,  -2.2627  ,
        -5.149   ,  -5.0636  ,   3.1405  ,   1.0793  ,  -0.72892 ,
        -3.9939  ,  -0.69551 ,  -0.55767 ,   3.2555  ,  -2.9449  ,
         4.7114  ,   1.6388  ,   1.3828  ,   1.4255  ,  -3.233

In [9]:
nlp(u'lion').vector.shape

(300,)

Como podemos observar, `lion` es representada por un vector de *300* dimensiones. Esto es lo que llamamos un `Word to Vector` o *Word2Vec*.

Sin embargo, esta no es la única característica de los modelos de Spacy. También podemos tener un vector de documento lo que significa que por cada documento vamos a promediar los vectores de cada vector que lo compone.

In [11]:
nlp(u'The quick brown fox jumped').vector.shape

(300,)

## Comprobando la similitud entre tokens

Recordemos que una de las características más importantes de estos vectores es que desbloquean exitosamente la semantica de las palabras con operaciones aritméticas, tarea que antes era muy difícil con las técnicas más clásicas.

Gracias a que ya contamos con vectores pre-computados, podemos medir la similitud de estas palabras. La similitud se mide con el coseno entre los vectores. 

El rango del coseno es $[-1,+1]$, donde valores cercanos a cero, indican que los vectores son ortogonales, es decir su ángulo es de 90 grados y se consideran disimilares. Cuando la similitud es positiva cercano a $1$, quiere decir que ambos vecrtores son similares, ya que su angulo es inferior a 90 grados. Finalmente, cuando la similitud es negativa, cercana a $-1$, quiere decir que los vectores son opuestos, su ángulo es mayor a 90 grados. Esto último es diferente a valores ceranos a 0 ya que lo opuesto también lleva consigo una relación semántica.

In [28]:
tokens = nlp('king queen jester')

Por ejemplo, podemos establecer una relación semántica entre un león y un gato, amobos son felinos. Además, entre un gato y una mascota, ya que un gato es usualmente una. Ahora verifiquemos esta suposición aritméticamente:

In [29]:
def print_similarity(tokens):
    for token in tokens:
        sim = {t:token.similarity(t) for t in tokens}
        print(f'Similarities with word {token}:\n{sim}\n')

print_similarity(tokens)

Similarities with word king:
{king: 1.0, queen: 0.6108841896057129, jester: 0.5094937682151794}

Similarities with word queen:
{king: 0.6108841896057129, queen: 1.0, jester: 0.35041603446006775}

Similarities with word jester:
{king: 0.5094937682151794, queen: 0.35041603446006775, jester: 1.0}



Naturalmente, cada palabra es completamente similar con sigo misma, por eso para lion, vemos que la similitud es 1. Nótese que la palabra león tiene una similitud de 0.5 con la palabra gato, pero al mismo tiempo tiene una baja similitud con la palabra mascota. 

Entonces verdaderamente la similitud entre los vectores puede revelar una relación semántica entre ellos. Y lo mejor es que es sistemáticamente computable con aritmética vectorial.

Observemos otros ejemplos:

In [30]:
tokens = nlp('like love hate')

Desde nuestro conocimeinto, sabemos que `love` (amor) y `hate` (odio) son antonimos, es decir, palabras con significado opuesto. Pero como ambas tienden a ser utilizadas en el mismo contexto, sus vectores pueden tener algo de similitud. Esto significa que los vectores son buenos para detectar contexto pero no para obtener definiciones u obtener un significado humanamente entendible de las palabras.

In [31]:
print_similarity(tokens)

Similarities with word like:
{like: 1.0, love: 0.5212638974189758, hate: 0.5065141320228577}

Similarities with word love:
{like: 0.5212638974189758, love: 1.0, hate: 0.5708349943161011}

Similarities with word hate:
{like: 0.5065141320228577, love: 0.5708349943161011, hate: 1.0}



Ahora, podemos inspeccionar el vocabulario de Spacy:

In [32]:
nlp.vocab.vectors.shape

(514157, 300)

Tenemos un total de $514,157$ palabras con $300$ dimensiones cada una.

Si una palabra no está presente en el vocabulario, significa que no va a tener un vector.

In [33]:
tokens = nlp('dog cat nargle')
for token in tokens:
    print(token.text, token.has_vector, token.vector_norm, token.is_oov)

dog True 75.254234 False
cat True 63.188496 False
nargle False 0.0 True


Esto debemos tenerlo en cuenta sobretodo a la hora de usar modelos pre-entrenados, si sabemos de antemano que nuestro corpus tiene tokens que no están en el vocabulario, pues lo más seguro es que nuestro modelo no sea lo suficientemente bueno.

## Calculando un nuevo vector
Técnicamente, podemos usar algebra lineal para obtener un vector de una nueva palabra. Veamos el ejemplo clásico de los word embeddings, la relación entre rey, hombre, mujer y reina:

In [34]:
from scipy import spatial

cosine_similarity = lambda v1, v2: 1 - spatial.distance.cosine(v1, v2)

In [35]:
king = nlp.vocab['king'].vector
man = nlp.vocab['man'].vector
woman = nlp.vocab['woman'].vector

Ahora podemos realizar operaciones simples como adición y substracción de vectores y ver lo que nos deja.

Si a `king` le restamos `man` y le sumamos `woman`, quizás nuestra intuición nos diga que el resultado debería ser la palabra `reina` verdad?

In [36]:
# Notice for king we substract the man features and add the women features.
# We expect the resulting vector to be similar to Queen, princess, highness, etc.
nv = king - man + woman

In [37]:
cosine_similarity(nv, nlp.vocab['queen'].vector)

0.6178014748458582

In [38]:
computed_similarities = {word.text:cosine_similarity(nv, word.vector) 
                         for word in nlp.vocab if word.has_vector and word.is_lower and word.is_alpha
                        }


In [39]:
computed_similarities = sorted(computed_similarities.items(), key=lambda kv: kv[1], reverse=True)

In [40]:
computed_similarities[:10]

[('king', 0.8489541040036218),
 ('queen', 0.6178014748458582),
 ('and', 0.38990046973701253),
 ('lord', 0.38926191190469983),
 ('that', 0.3848358043284852),
 ('jester', 0.37551937617847275),
 ('tyrant', 0.3478205322350596),
 ('where', 0.3385923464745617),
 ('president', 0.3376784032992368),
 ('democracy', 0.3343139683773124)]

Observemos la medida se similitud para la palabra `queen`, is significativamente cerca a lo que queríamos!